src/json2typescript/json-convert.ts

```diff
@@ -669,6 +763,12 @@ export class JsonConvert {
         // Map the property
         try {
             json[jsonPropertyName] = customConverter !== null ? customConverter.serialize(classInstancePropertyValue) : this.verifyProperty(expectedJsonType, classInstancePropertyValue, true);
+
+            const classConstructorName = dataObject?.constructor?.name;
+
+            if (this._useDiscriminator && this._classes[classConstructorName]) {
+                json[this._discriminatorPropertyName] = classConstructorName;
+            }
```

Comment:
I tested your changes and I found a potential issue here.

The local (TypeScript) constructor name (for example `Dog`) does not necessarily match the `$type` property coming from the server. For example, the server property `$type` could be defined as string `"DogDto"`.

Now imagine your local `classes` are defined as `classes = { "DogDto": Dog }`. In this case, your deserialization works as expected, but the serialization here fails. When doing serialization, we should not set `$type` to `Dog` (this is what the code is doing as of now), but it should be set again to `DogDto`! Basically, we need to compare the `classConstructorName` to the constructor names from all values (not keys!) of the `classes` data.

Your code works in the assumption that `$type` from the backend exactly matches the class name in the frontend. Is this a necessary convention? In any case, I think we should improve this. If you agree, I will add a commit.

Reply:
Assuming `this.classes` is now a `Map`, then this is my suggestion for the serialization:

```
const classConstructorName = dataObject?.constructor?.name;

if (this._useDiscriminator) {
    this.classes.forEach((classDataObject: {new(): any}, key: string) => {
        if (classDataObject.name === classConstructorName) {
             json[this._discriminatorPropertyName] = key; // use the Map's key and NOT the classConstructorName!
         }
    });
}
```

Reply:
Thanks for testing the code and the great feedback!
Of course it makes a lot more sense to use the name provided in the list instead of the constructor name.
I just didn't think of this, since I always used the same string for the name and the class itself.

Your solution looks fine, so please push it :smiley: